### PR TITLE
feat: multi-treasury transfer contract RNSCommission

### PR DESCRIPTION
**Pull request by tringuyenskymavis**
### Description
This PR creates an RNSCommission contract for multi-treasury transfer.

### CODE REVIEW

File: src/RNSCommission.sol [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    if (sum != MAX_PERCENTAGE) revert InvalidRatio();
+
+    emit TreasuriesUpdated(treasuriesInfo);
+  }
+
+  /// Check if `sender` is allowed to send money
+  function _isAllowedSender(address sender) internal view returns (bool) {
+    return _allowedSenders[sender];
+  }
```
Comment (ducthotran2010) [inline src/RNSCommission.sol]: Can we use `_allowedSenders[sender]` directly?

File: src/RNSCommission.sol [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  mapping(address => bool) public _allowedSenders;
+
+  constructor() {
+    _disableInitializers();
+  }
+
+  receive() external payable {
+    if (_isAllowedSender(msg.sender)) {
+      _allocateCommissionAndTransferToTreasury(msg.value);
+    }
```
Comment (ducthotran2010) [inline src/RNSCommission.sol]: Don't repeat yourself: https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

File: src/RNSCommission.sol [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  fallback() external payable {
+    if (_isAllowedSender(msg.sender)) {
+      _allocateCommissionAndTransferToTreasury(msg.value);
+    }
+  }
```
Comment (TuDo1403) [inline src/RNSCommission.sol]: `fallback` is not needed.

File: src/RNSCommission.sol [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+contract RNSCommission is Initializable, AccessControlEnumerable, INSCommission {
+  uint256 public constant MAX_PERCENTAGE = 100_00;
+  bytes32 public constant COMMISSION_SETTER_ROLE = keccak256("COMMISSION_SETTER_ROLE");
+
+  /// @dev Gap for upgradability.
+  uint256[50] private ____gap;
+
+  Commission[] internal _commissionInfo;
+  mapping(address => bool) public _allowedSenders;
```
Comment (TuDo1403) [inline src/RNSCommission.sol]:
```suggestion
  mapping(address sender => bool allowed) public _allowedSenders;
```

File: src/RNSCommission.sol [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    uint256 sendersLength = allowedSenders.length;
+    for (uint256 i; i < sendersLength; ++i) {
+      _allowedSenders[allowedSenders[i]] = true;
+    }
+
+    _setTreasuries(treasuryCommission);
+  }
+
+  /// @inheritdoc INSCommission
+  function getTreasuries() external view returns (Commission[] memory treasuriesInfo) {
```
Comment (ducthotran2010) [inline src/RNSCommission.sol]:
```suggestion
  function getCommissions() external view returns (Commission[] memory treasuriesInfo) {
```

File: src/RNSCommission.sol [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+    _commissionInfo[treasuryId].recipient = newAddr;
+    _commissionInfo[treasuryId].name = name;
+    emit TreasuryInfoUpdated(newAddr, name, treasuryId);
+  }
+
+  /// @inheritdoc INSCommission
+  function allowSender(address sender) external onlyRole(COMMISSION_SETTER_ROLE) {
+    _allowedSenders[sender] = true;
+  }
```
Comment (TuDo1403) [inline src/RNSCommission.sol]: how can we unset specific sender?

File: src/RNSCommission.sol [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+import { Initializable } from "@openzeppelin/contracts/proxy/utils/Initializable.sol";
+import { AccessControlEnumerable } from "@openzeppelin/contracts/access/AccessControlEnumerable.sol";
+import { INSCommission } from "./interfaces/INSCommission.sol";
+import { Math } from "@openzeppelin/contracts/utils/math/Math.sol";
+import { RONTransferHelper } from "./libraries/transfers/RONTransferHelper.sol";
+
+contract RNSCommission is Initializable, AccessControlEnumerable, INSCommission {
+  uint256 public constant MAX_PERCENTAGE = 100_00;
+  bytes32 public constant COMMISSION_SETTER_ROLE = keccak256("COMMISSION_SETTER_ROLE");
```
Comment (TuDo1403) [inline src/RNSCommission.sol]: missing doc

File: src/RNSCommission.sol [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    return _commissionInfo;
+  }
+
+  /// @inheritdoc INSCommission
+  function setTreasuries(Commission[] calldata treasuriesInfo) external onlyRole(COMMISSION_SETTER_ROLE) {
+    _setTreasuries(treasuriesInfo);
+  }
+
+  /// @inheritdoc INSCommission
+  function changeTreasuryInfo(address payable newAddr, bytes calldata name, uint256 treasuryId)
```
Comment (ducthotran2010) [inline src/RNSCommission.sol]:
```suggestion
  function changeTreasuryInfo(uint256 treasuryId, address payable newAddr, bytes calldata name)
```

File: src/interfaces/INSCommission.sol [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+  struct Allocation {
+    address payable recipient;
+    uint256 value;
+  }
+
+  /// @dev Emitted when all the treasury info info are updated.
+  event TreasuriesUpdated(Commission[] treasuriesInfo);
+  /// @dev Emitted when specific treasury info are updated.
+  event TreasuryInfoUpdated(address payable treasuryAddr, bytes name, uint256 treasuryId);
```
Comment (TuDo1403) [inline src/interfaces/INSCommission.sol]:
```suggestion
  /// @dev Emitted when all the treasury info info are updated.
  event TreasuriesUpdated(address indexed updatedBy, Commission[] treasuriesInfo);
  /// @dev Emitted when specific treasury info are updated.
  event TreasuryInfoUpdated(address indexed updatedBy, address payable treasuryAddr, bytes name, uint256 treasuryId);
```

File: src/RNSCommission.sol [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+  Commission[] internal _commissionInfo;
+  mapping(address => bool) public _allowedSenders;
+
+  constructor() {
+    _disableInitializers();
+  }
+
+  receive() external payable {
+    if (_isAllowedSender(msg.sender)) {
```
Comment (TuDo1403) [inline src/RNSCommission.sol]: what if sender is not allowed?

File: src/RNSCommission.sol [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+  function setTreasuries(Commission[] calldata treasuriesInfo) external onlyRole(COMMISSION_SETTER_ROLE) {
+    _setTreasuries(treasuriesInfo);
+  }
+
+  /// @inheritdoc INSCommission
+  function changeTreasuryInfo(address payable newAddr, bytes calldata name, uint256 treasuryId)
+    external
+    onlyRole(COMMISSION_SETTER_ROLE)
+  {
+    if (treasuryId < 0 || treasuryId > _commissionInfo.length - 1) {
```
Comment (TuDo1403) [inline src/RNSCommission.sol]: `treasuryId` never < 0

**Review by TuDo1403 — CHANGES_REQUESTED** [review]

File: src/interfaces/INSCommission.sol [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+  function setTreasuries(Commission[] calldata treasuriesInfo) external;
+
+  /**
+   * @dev Sets for specific treasury information based on the treasury `id`.
+   *
+   * Requirements:
+   * - The method caller is setter role.
+   * Emits the event `TreasuryInfoUpdated`.
+   */
+  function changeTreasuryInfo(address payable newAddr, bytes calldata name, uint256 treasuryId) external;
```
Comment (ducthotran2010) [inline src/interfaces/INSCommission.sol]: - [change](https://www.etymonline.com/search?q=change): to make (something) other than what it was, cause to turn or pass from one state to another
- [set](https://www.etymonline.com/search?q=set): put in a definite place
```suggestion
  function setTreasuryInfo(uint256 treasuryId, address payable newAddr, bytes calldata name) external;
```

File: src/RNSCommission.sol [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+import { RONTransferHelper } from "./libraries/transfers/RONTransferHelper.sol";
+
+contract RNSCommission is Initializable, AccessControlEnumerable, INSCommission {
+  uint256 public constant MAX_PERCENTAGE = 100_00;
+  bytes32 public constant COMMISSION_SETTER_ROLE = keccak256("COMMISSION_SETTER_ROLE");
+
+  /// @dev Gap for upgradability.
+  uint256[50] private ____gap;
+
+  Commission[] internal _commissionInfo;
```
Comment (TuDo1403) [inline src/RNSCommission.sol]:
```suggestion
  Commission[] internal _commissionInfos;
```

File: src/RNSCommission.sol (hunk 4 above)
Comment (TuDo1403) [inline src/RNSCommission.sol]:
```suggestion
  mapping(address => bool) public _allowedSender;
```

**Review by TuDo1403 — CHANGES_REQUESTED** [review]

File: src/RNSCommission.sol (hunk 11 above)
Comment (ducthotran2010) [inline src/RNSCommission.sol]: Since treasuryId is uint256, it cannot be < 0

File: src/RNSCommission.sol [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  function _setTreasuries(Commission[] calldata treasuriesInfo) internal {
+    uint256 length = treasuriesInfo.length;
+    // treasuriesInfo[] can not be empty
+    if (length < 1) revert InvalidArrayLength();
+
+    delete _commissionInfo;
+
+    uint256 sum = 0;
```
Comment (TuDo1403) [inline src/RNSCommission.sol]:
```suggestion
    uint256 sum;
```

File: src/RNSCommission.sol [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+      allocs[i] = Allocation({
+        recipient: _commissionInfo[i].recipient,
+        value: _computePercentage(totalAmount, _commissionInfo[i].ratio)
+      });
+      sumValue += allocs[i].value;
+    }
+
+    // Refund the remaining RON to the last treasury
+    if (sumValue < totalAmount) {
+      allocs[lastIdx] = Allocation({ recipient: _commissionInfo[lastIdx].recipient, value: totalAmount - sumValue });
```
Comment (ducthotran2010) [inline src/RNSCommission.sol]: This code replaces value of the last recipient

**Review by ducthotran2010 — CHANGES_REQUESTED** [review]
Consider fixing the issues I mentioned in the last PR

File: src/interfaces/INSCommission.sol (hunk 9 above)
Comment (ducthotran2010) [reply src/interfaces/INSCommission.sol]: `uint256 indexed treasuryId`

File: src/RNSCommission.sol [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+      address payable recipient = allocs[i].recipient;
+
+      RONTransferHelper.safeTransfer(recipient, value);
+    }
+  }
+
+  function _setTreasuries(Commission[] calldata treasuriesInfo) internal {
+    uint256 length = treasuriesInfo.length;
+    // treasuriesInfo[] can not be empty
+    if (length < 1) revert InvalidArrayLength();
```
Comment (TuDo1403) [inline src/RNSCommission.sol]:
```suggestion
    if (length == 0) revert InvalidArrayLength();
```

File: src/RNSCommission.sol [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+
+    delete _commissionInfos;
+
+    uint256 sum;
+
+    for (uint256 i = 0; i < length; ++i) {
```
Comment (TuDo1403) [inline src/RNSCommission.sol]:
```suggestion
    for (uint256 i; i < length; ++i) {
```

File: src/RNSCommission.sol [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  /**
+   * @dev Helper method to allocate commission and take fee into treasuries address.
+   */
+  function _allocateCommissionAndTransferToTreasury(uint256 ronAmount) internal {
+    INSCommission.Allocation[] memory allocs = _calcAllocations(ronAmount);
+    uint256 length = allocs.length;
+
+    for (uint256 i = 0; i < length; ++i) {
```
Comment (TuDo1403) [inline src/RNSCommission.sol]: v

File: src/interfaces/INSCommission.sol [hunk 19]
```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+interface INSCommission {
+  struct Commission {
+    address payable recipient;
+    uint256 ratio; // Values [0; 100_00] reflexes [0; 100%]
+    bytes name; // Treasury's name
```
Comment (TuDo1403) [inline src/interfaces/INSCommission.sol]:
```suggestion
    bytes name; // Treasury's name
```
not need

**Review by nxqbao — DISMISSED** [review]

**Review by ducthotran2010 — DISMISSED** [review]
minor:
* treasury name
* commission <> treasury

File: src/RNSCommission.sol [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+  uint256 public constant MAX_PERCENTAGE = 100_00;
+  /// @dev Role for accounts that can set commissions infomation and grant or revoke `SENDER_ROLE`.
+  bytes32 public constant COMMISSION_SETTER_ROLE = keccak256("COMMISSION_SETTER_ROLE");
+  /// @dev Role for accounts that can send RON for this contract.
+  bytes32 public constant SENDER_ROLE = keccak256("SENDER_ROLE");
+
+  /// @dev Gap for upgradability.
+  uint256[50] private ____gap;
+
+  Commission[] internal _commissionInfos;
```
Comment (TuDo1403) [inline src/RNSCommission.sol]: should provide doc

File: src/RNSCommission.sol [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+    emit CommissionInfoUpdated(msg.sender, commissionIdx, newRecipient, newName);
+  }
+
+  /**
+   * @dev Helper method to calculate allocations.
+   */
+  function _calcAllocations(uint256 totalAmount) internal view returns (Allocation[] memory allocs) {
+    if (totalAmount == 0) {
+      revert InvalidAmountOfRON();
+    }
```
Comment (TuDo1403) [inline src/RNSCommission.sol]:
```suggestion
    if (totalAmount == 0) revert InvalidAmountOfRON();
```

File: src/RNSCommission.sol [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+  /**
+   * @dev Helper method to allocate commission and take fee into recipient address.
+   */
+  function _allocateCommissionAndTransferToRecipient(uint256 ronAmount) internal {
+    INSCommission.Allocation[] memory allocs = _calcAllocations(ronAmount);
+    uint256 length = allocs.length;
+
+    for (uint256 i; i < length; ++i) {
+      uint256 value = allocs[i].value;
+      address payable recipient = allocs[i].recipient;
```
Comment (TuDo1403) [inline src/RNSCommission.sol]: I don't see why we should cache those value to stack.

File: src/interfaces/INSCommission.sol [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+interface INSCommission {
+  struct Commission {
+    address payable recipient;
+    uint256 ratio; // Values [0; 100_00] reflexes [0; 100%]
+    string name; // Commission's name
+  }
+
+  struct Allocation {
+    address payable recipient;
+    uint256 value;
```
Comment (TuDo1403) [inline src/interfaces/INSCommission.sol]: this is not `value`. Renamed to something else, for example `feeBps`

**Review by TuDo1403 — CHANGES_REQUESTED** [review]

File: src/RNSCommission.sol [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+    for (uint256 i; i < length; ++i) {
+      _setupRole(COMMISSION_SETTER_ROLE, commissionSetters[i]);
+    }
+
+    uint256 sendersLength = allowedSenders.length;
+    for (uint256 i; i < sendersLength; ++i) {
+      _setupRole(SENDER_ROLE, allowedSenders[i]);
+    }
+
+    _setRoleAdmin(SENDER_ROLE, COMMISSION_SETTER_ROLE);
```
Comment (TuDo1403) [inline src/RNSCommission.sol]: this is not needed.

File: src/RNSCommission.sol [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+  function setCommissions(Commission[] calldata commissionInfos) external onlyRole(COMMISSION_SETTER_ROLE) {
+    _setCommissions(commissionInfos);
+  }
+
+  /// @inheritdoc INSCommission
+  function setCommissionInfo(uint256 commissionIdx, address payable newRecipient, string calldata newName)
+    external
+    onlyRole(COMMISSION_SETTER_ROLE)
+  {
+    if (commissionIdx > _commissionInfos.length - 1) {
```
Comment (TuDo1403) [inline src/RNSCommission.sol]:
```suggestion
    if (commissionIdx >= _commissionInfos.length) {
```

File: src/RNSCommission.sol [hunk 26]
```diff
@@ -0,0 +1,148 @@
...
+import { Initializable } from "@openzeppelin/contracts/proxy/utils/Initializable.sol";
+import { AccessControlEnumerable } from "@openzeppelin/contracts/access/AccessControlEnumerable.sol";
+import { INSCommission } from "./interfaces/INSCommission.sol";
+import { RONTransferHelper } from "./libraries/transfers/RONTransferHelper.sol";
+
+contract RNSCommission is Initializable, AccessControlEnumerable, INSCommission {
+  /// @dev Constant representing the maximum percentage value (100%).
+  uint256 public constant MAX_PERCENTAGE = 100_00;
+  /// @dev Role for accounts that can set commissions infomation and grant or revoke `SENDER_ROLE`.
+  bytes32 public constant COMMISSION_SETTER_ROLE = keccak256("COMMISSION_SETTER_ROLE");
```
Comment (TuDo1403) [inline src/RNSCommission.sol]: I don't think `COMMISSION_SETTER_ROLE` is needed.

File: src/RNSCommission.sol [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+   */
+  function _allocateCommissionAndTransferToRecipient(uint256 ronAmount) internal {
+    INSCommission.Allocation[] memory allocs = _calcAllocations(ronAmount);
+    uint256 length = allocs.length;
+
+    for (uint256 i; i < length; ++i) {
+      uint256 value = allocs[i].value;
+      address payable recipient = allocs[i].recipient;
+
+      RONTransferHelper.safeTransfer(recipient, value);
```
Comment (TuDo1403) [inline src/RNSCommission.sol]: should emit event when allocate to treasury

File: src/RNSCommission.sol [hunk 28]
```diff
@@ -0,0 +1,148 @@
...
+import { INSCommission } from "./interfaces/INSCommission.sol";
+import { RONTransferHelper } from "./libraries/transfers/RONTransferHelper.sol";
+
+contract RNSCommission is Initializable, AccessControlEnumerable, INSCommission {
+  /// @dev Constant representing the maximum percentage value (100%).
+  uint256 public constant MAX_PERCENTAGE = 100_00;
+  /// @dev Role for accounts that can set commissions infomation and grant or revoke `SENDER_ROLE`.
+  bytes32 public constant COMMISSION_SETTER_ROLE = keccak256("COMMISSION_SETTER_ROLE");
+  /// @dev Role for accounts that can send RON for this contract.
+  bytes32 public constant SENDER_ROLE = keccak256("SENDER_ROLE");
```
Comment (TuDo1403) [inline src/RNSCommission.sol]: this also should have declarations in interface.

File: src/interfaces/INSCommission.sol [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  /// @dev Emitted when all the commission info is updated.
+  event CommissionsUpdated(address indexed updatedBy, Commission[] commissionInfos);
+  /// @dev Emitted when specific commission info is updated.
+  event CommissionInfoUpdated(
+    address indexed updatedBy, uint256 indexed commissionIdx, address payable newRecipient, string newName
+  );
+  /// @dev Emiited when transfer RON to commission's recipient.
+  event Transfer(address indexed recipient, uint256 commissionAmount);
```
Comment (TuDo1403) [inline src/interfaces/INSCommission.sol]:
```suggestion
  event Distributed(address indexed recipient, uint256 commissionAmount);
```

**Review by TuDo1403 — DISMISSED** [review]
LGTM. Please make sure the logic is propoperly tested.

**Review by TuDo1403 — APPROVED** [review]

**Comment by TuDo1403** [comment]
Closed in favor of #220

**Pull request closed without merging**